src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,168 @@
+            ImmutableArray<ReferenceInfo>.Builder unusedReferences)
+        {
+            var usedAssemblyLookup = usedCompilationAssemblies.ToImmutableHashSet();
+
+            // Determine which toplevel references have compilation assemblies in set of used assemblies.
+            var toplevelUsedReferences = references.Where(reference =>
+                reference.CompilationAssemblies.Any(usedAssemblyLookup.Contains));
+            var toplevelUsedAssemblyReferences = toplevelUsedReferences.SelectMany(reference => reference.GetAllCompilationAssemblies());
+
+            // Remove all assemblies that are brought into the compilation from those toplevel used references. When
+            // determining transtively used references this will reduce false positives.
+            var remainingReferences = references.Except(toplevelUsedReferences);
+            var remainingUsedAssemblyReferences = usedCompilationAssemblies.Except(toplevelUsedAssemblyReferences);
+            var remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which transtive references have compilation assemblies in the set of remaining used assemblies.
+            foreach (var reference in remainingReferences)
+            {
+                var allCompilationAssemblies = reference.GetAllCompilationAssemblies();
+                if (allCompilationAssemblies.IsEmpty)
+                {
+                    // We will consider References that do not contribute any assemblies to the compilation,
+                    // such as Analyzer packages, as used.
+                    continue;
+                }
+
+                if (!allCompilationAssemblies.Any(remainingUsedAssemblyLookup.Contains))
+                {
+                    // None of the assemblies brought into this compilation are in the remaining
+                    // used assemblies list, so we will consider the reference unused.
+                    unusedReferences.Add(reference);
+                    continue;
+                }
+
+                remainingUsedAssemblyReferences = remainingUsedAssemblyReferences.Except(allCompilationAssemblies);
+                remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+            }
+
+            return remainingUsedAssemblyReferences;
+        }
+
+        public async Task<Project> UpdateReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken)
+        {
+            var referenceCleanupService = project.Solution.Workspace.Services.GetRequiredService<IReferenceCleanupService>();
+
+            await ApplyReferenceUpdatesAsync(referenceCleanupService, project.FilePath!, referenceUpdates, cancellationToken).ConfigureAwait(false);
+
+            return project.Solution.Workspace.CurrentSolution.GetProject(project.Id)!;
```

Comment:
Do we have a `GetRequiredProject` that could be used here?

Reply:
Oh hey an extension method. Updated.